pkg/azuredx/azureauth/scopes.go

```diff
@@ -0,0 +1,35 @@
+package azureauth
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana-azure-sdk-go/azcredentials"
+	"github.com/grafana/grafana-azure-sdk-go/azsettings"
+)
+
+var (
+	azureDataExplorerScopes = map[string][]string{
+		azsettings.AzurePublic:       {"https://kusto.kusto.windows.net/.default"},
+		azsettings.AzureChina:        {"https://kusto.kusto.chinacloudapi.cn/.default"},
+		azsettings.AzureUSGovernment: {"{clusterUrl}/.default"},
+	}
+)
+
+func getAzureScopes(credentials *azcredentials.AzureClientSecretCredentials, clusterUrl string) ([]string, error) {
+	azureCloud := credentials.AzureCloud
+
+	// Get scopes for the given cloud
+	if scopesTmpl, ok := azureDataExplorerScopes[azureCloud]; !ok {
+		err := fmt.Errorf("the Azure cloud '%s' not supported by Azure Data Explorer datasource", azureCloud)
+		return nil, err
+	} else {
+		// TODO: #356 Generalize generation of scopes for all clouds
+		clusterUrl = strings.TrimSuffix(clusterUrl, "/")
+		scopes := make([]string, len(scopesTmpl))
+		for i, scopeTmpl := range scopesTmpl {
+			scopes[i] = strings.Replace(scopeTmpl, "{clusterUrl}", clusterUrl, 1)
```

Comment:
> While in this particular case ADX endpoint needs only a single scope (`.default` - is the default, catch-all scope), generally it's not always the case. An endpoint in theory _may_ require more than one scope at the same time, and access token can be issued for more than one scope.
> 
> Signature of the `GetAccessToken` func in `AzureTokenProvider` ([here](https://github.com/grafana/grafana-azure-sdk-go/blob/e079ca81c3fb6b309b37843345cf97f30aad4609/aztokenprovider/token_provider.go#L20)):
> 
> ```go
> type AzureTokenProvider interface {
> 	GetAccessToken(ctx context.Context, scopes []string) (string, error)
> }
> ```
> 
> So I made `getAzureScopes` to return `[]string` array more for convention, as a type for representation of scopes, proper abstraction, rather than because it's technically required. The fact that `getAzureScopes` returns only one scope is implementation detail.



If ADX only ever will return one scope, just for this ADX implementation, what do you think if we returned only `string` and reduced the complexity from the loop here?
(or return the one string wrapped in `[]string{}`)

something along the lines of:

```
var (
	azureDataExplorerScopes = map[string]string{
		azsettings.AzurePublic:       "https://kusto.kusto.windows.net/.default",
		azsettings.AzureChina:        "https://kusto.kusto.chinacloudapi.cn/.default",
		azsettings.AzureUSGovernment: "{clusterUrl}/.default",
	}
)

func getAzureScopes(credentials *azcredentials.AzureClientSecretCredentials, clusterUrl string) ([]string, error) {
	azureCloud := credentials.AzureCloud

	// Get scopes for the given cloud
	scopeTmpl, ok := azureDataExplorerScopes[azureCloud]
	if !ok {
		return nil, fmt.Errorf("the Azure cloud '%s' not supported by Azure Data Explorer datasource", azureCloud)
	}

	return []string{strings.Replace(scopeTmpl, "{clusterUrl}", strings.TrimSuffix(clusterUrl, "/"), 1)}, nil
}

```

Reply:
Yes, makes sense to just wrap the string. I will update the code.

Reply:
Fixed